src/Server.cpp

```diff
@@ -928,3 +965,67 @@ void cleanupClosedConnections(Server& server)
 			++iter;
 	}
 }
+
+/**
+ * @brief Iterates through all connections, closes and removes idle ones.
+ *
+ * The for loop through the connections map has no increment statement because the
+ * iterator is incremented in the loop body. If .erase() is called on an iterator,
+ * it is invalidated.
+ *
+ * @param server The server object to cleanup idle connections for.
+ */
+void cleanupIdleConnections(Server& server)
+{
+	for (std::map<int, Connection>::iterator iter = server.getConnections().begin();
+		 iter != server.getConnections().end();
+		/* no iter*/) {
+		if (iter->second.m_status == Connection::Idle) {
+			close(iter->first);
+			server.getConnections().erase(iter++);
+		} else
+			++iter;
+	}
+}
+
+/**
+ * @brief Performs a gracefule shutdown of the server.
+ *
+ * Closes all virtual servers, cleans up idle and closed connections. Then enters another event loop with an additional
+ * condition: as long as active connections exist. This second loop always cleans up idle connections. It can also
+ * be interrupted with another signal, which aborts the graceful shutdown.
+ *
+ * @param server The server object to shut down.
+ * @sa https://pkg.go.dev/net/http#Server.Shutdown
+ */
+void serverShutdown(Server& server)
+{
+	LOG_DEBUG << "Closing all virtual servers";
+	for (std::map<int, Socket>::iterator iter = server.getVirtualServers().begin();
+		 iter != server.getVirtualServers().end(); ++iter) {
+		server.removeEvent(iter->first);
+		close(iter->first);
+	}
+	server.getVirtualServers().clear();
+
+	LOG_DEBUG << "Cleanup idle connections";
+	cleanupIdleConnections(server);
+	cleanupClosedConnections(server);
+
+	LOG_DEBUG << "Waiting for connections to finish";
+	while (g_signalStatus == 0 && !server.getConnections().empty()) {
```

Comment:
We can process them - the server itself is running

The virtual servers are our listening points. By shutting them down we don't accept any new connections.
Then we handle connections which are already in the middle of processing (e.g we received a GET request and just need to send a response). 
This is called graceful shutdown - mimics implementation of go: https://pkg.go.dev/net/http#Server.Shutdown

We can interrupt the graceful shutdown by sending another signal, which would break out of the second loop, which handles the remaining connections. If a client sends us a big file, we can stop it and exit non graceful

Reply:
Thanks for clearing that up